### PR TITLE
Use real constructors

### PR DESCRIPTION
**Pull request by rakina**
For  #46, #56, #57, and #58. If there's still some unresolved discussion on those threads and this change is invalid, please ignore. 

cc @domenic @bzbarsky (somehow I can't add any of you to the "Reviewers" list...)

***
<a href="https://pr-preview.s3.amazonaws.com/WICG/construct-stylesheets/pull/61.html" title="Last updated on Nov 15, 2018, 2:07 AM GMT (b261ef7)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/WICG/construct-stylesheets/61/ea0dcc9...b261ef7.html" title="Last updated on Nov 15, 2018, 2:07 AM GMT (b261ef7)">Diff</a>

### CODE REVIEW

**Review by kamilkem — APPROVED** [review]

File: index.bs [hunk 1]
```diff
@@ -47,66 +51,88 @@ dictionary CSSStyleSheetInit {
...
+	<dt><dfn  for=CSSStyleSheet lt="constructed flag">constructed flag</dfn></dt>
+	<dd>
+		Specified when created. Either set or unset. Unset by default.
+		Signifies whether this stylesheet is made via constructor or not, so must be set only for stylesheets that are constructed using the {{CSSStyleSheet()}} or {{CSSStyleSheet(options)}} function.
+	</dd>
+	<dt><dfn  for=CSSStyleSheet lt="disallow modification flag">disallow modification flag</dfn></dt>
+	<dd>
+		Either set or unset. Unset by default. If set, modification to the stylesheet's rules are not allowed.
+	</dd>
+	<dt><dfn  for=CSSStyleSheet lt="constructor document">constructor document</dfn></dt>
```
Comment (domenic) [inline index.bs]: What are constructor document and set of adopter documents used for? They are set, but nobody ever reads them. I was hoping we wouldn't have to maintain this information, to be honest.

File: index.bs [hunk 2]
```diff
@@ -139,6 +165,8 @@ Every {{DocumentOrShadowRoot}} has <dfn>adopted stylesheets</dfn>.
 The user agent must include all style sheets in the {{DocumentOrShadowRoot}}'s
 [=adopted stylesheets=] inside its <a>document or shadow root CSS style sheets</a>. The [=adopted stylesheets=] are ordered after all the other style sheets (i.e. those derived from {{DocumentOrShadowRoot/styleSheets}}).
 
+If a {{DocumentOrShadowRoot}} is moved into a different {{Document}}, its [=adopted stylesheets=] are also moved along with it, retaining the same set of style sheets.
```
Comment (domenic) [inline index.bs]: This statement is unclear to me. Is this a statement of fact that follows from another part of the spec, and is trying to summarize things in an easy to understand way? If so, I am having a hard time understanding it. Or is this trying to impose a normative implementation requirement? If so, I'm unsure what parts of the code would need to be modified to make this true.

File: index.bs [hunk 3]
```diff
@@ -47,66 +51,88 @@ dictionary CSSStyleSheetInit {
...
+	<dt><dfn  for=CSSStyleSheet lt="set of adopter documents">set of adopter documents</dfn></dt>
+	<dd>
+		A set of {{Document}}s on which the stylesheet is adopted (e.g. there is at least one {{DocumentOrShadowRoot}} within that {{Document}}'s tree on which the sheet is part of the {{DocumentOrShadowRoot}}'s {{DocumentOrShadowRoot/adoptedStyleSheets}}).
+	</dd>
+</dl>
+
+Modifying Constructed Stylesheets {#modifying-constructed-stylesheets}
+=============================
+
+After construction, constructed stylesheets can be modified using rule modification methods like <a spec="cssom-1" for="CSSStyleSheet">insertRule(rule)</a> or <a spec="cssom-1" for="CSSStyleSheet">deleteRule(rule)</a>, or {{replace(text)}} and {{replaceSync(text)}} if the sheet's [=disallow modification flag=] is not set. If those methods are called when the sheet's [=disallow modification flag=] is set, or <a spec="cssom-1" for="CSSStyleSheet">insertRule(rule)</a> is used to add an <a spec=css-cascade-4>@import</a> rule, a "{{NotAllowedError}}" {{DOMException}} must be thrown.
```
Comment (domenic) [inline index.bs]: This is OK, but it's a classic [monkey patch](https://annevankesteren.nl/2014/02/monkey-patch). As such we should make it clearer exactly what the patch is, that you would eventually change the CSSOM spec for. I'd suggest something like the following, although more precise and using the correct spec terms:

insertRule(rule)

1. If this sheet's disallow modification flag is set, throw....
1. If rule is an import rule and this sheet's constructed flag is set, throw...
1. (The rest of the algorithm remains as in CSSOM)

and the same for all other methods that are affected.

File: index.bs [hunk 4]
```diff
@@ -47,66 +51,88 @@ dictionary CSSStyleSheetInit {
...
+
+After construction, constructed stylesheets can be modified using rule modification methods like <a spec="cssom-1" for="CSSStyleSheet">insertRule(rule)</a> or <a spec="cssom-1" for="CSSStyleSheet">deleteRule(rule)</a>, or {{replace(text)}} and {{replaceSync(text)}} if the sheet's [=disallow modification flag=] is not set. If those methods are called when the sheet's [=disallow modification flag=] is set, or <a spec="cssom-1" for="CSSStyleSheet">insertRule(rule)</a> is used to add an <a spec=css-cascade-4>@import</a> rule, a "{{NotAllowedError}}" {{DOMException}} must be thrown. 
+
+<dl>
+    <dt><dfn method for=CSSStyleSheet lt="replace(text)">replace(text)</dfn></dt>
+    <dd>
+		1. Let <var>sheet</var> be the stylesheet on which this function is called on.
+		2. If <var>sheet’s</var> [=constructed flag=] is not set, or <var>sheet's</var> [=disallow modification flag=] is set, throw a "{{NotAllowedError}}" {{DOMException}}.
+		3. Set <var>sheet’s</var> CSS rules to an empty list, and set <var>sheet's</var> [=disallow modification flag=].
+		4. <a spec=css-syntax-3>Parse a stylesheet</a> from {{text}}.
```
Comment (domenic) [inline index.bs]: It looks like this doesn't return a list of rules; instead it returns a stylesheet of some sort. (But probably not the same as a CSSOM stylesheet?) This will probably need to be refactored eventually, but that's more of a @tabatkins issue. It doesn't need to block this, but we may want to file a tracking issue for this.

File: index.bs [hunk 5]
```diff
@@ -47,66 +51,88 @@ dictionary CSSStyleSheetInit {
...
 			If it returned a list of rules,
 			assign the list as <var>sheet’s</var> CSS rules;
 			otherwise,
 			set <var>sheet’s</var> CSS rules to an empty list.
```
Comment (domenic) [inline index.bs]: Preexisting editorial issue: every time you reference a "field" of the stylesheet concept, you should be cross-referencing it. E.g. `[=CSSStyleSheet/CSS rules=]`. But, I guess these are currently not exported, so that won't work... something to fix when we do the eventual integration into CSSOM.

File: index.bs (hunk 2 above)
Comment (rakina) [reply index.bs]: Hmm yeah I'm not sure about this too, it's to resolve #46. Do you have a better place and wording for it?

Comment (domenic) [reply index.bs]: Well, how are you planning on implementing this?

File: index.bs (hunk 1 above)
Comment (rakina) [reply index.bs]: It's used in line 114 (it's a really long line so you can't see it unless you scroll to the right, oops):
> If <var>sheet</var> contains one or more <a spec=css-cascade-4>@import</a> rules and <var>sheet's</var> [=set of adopter documents=] contains any other {{Document}} than <var>sheet's</var> [=constructor document=]

We don't want to allow `replace` calls with import rules, if it's already used in another document. 
See https://github.com/WICG/construct-stylesheets/issues/57#issuecomment-432644836 and https://github.com/WICG/construct-stylesheets/issues/57#issuecomment-436509375

File: index.bs (hunk 2 above)
Comment (rakina) [reply index.bs]: On the implementation side, it's just automatically carried over along with the `DocumentOrShadowRoot` since it's the same object, and the `CSSStyleSheet` objects are not changed in any way, the base URLs stay the same, the @imports are not re-resolved, etc. 
 
Maybe we should modify "adopt a node" spec as well?

File: index.bs (hunk 4 above)
Comment (rakina) [reply index.bs]: Looks like there's https://drafts.csswg.org/css-syntax-3/#parse-list-of-rules, used that now.

Comment (domenic) [reply index.bs]: That changes the handling of initial charset rules, but I am unsure whether it does so in a good or bad way.

File: index.bs (hunk 5 above)
Comment (rakina) [reply index.bs]: Added cross references.

File: index.bs (hunk 3 above)
Comment (rakina) [reply index.bs]: I see, added those. Should I keep these sentences as summary?

File: index.bs (hunk 4 above)
Comment (tabatkins) [reply index.bs]: That @charset handling is just so that other rules that check for "is this the first rule in the stylesheet?" don't fail because of a charset indicator.

That said, you bring up a good point - I should make the parsing algorithms *optionally* just allow themselves to be parsed in a forced utf-8, and then we should force that here. No reason to support other charsets; that's a legacy thing.

The *other* difference is whether `<!--` and `-->` are ignored during parsing or not. This is another *very* legacy thing, to support commenting out the contents of `style` blocks to hide them from browsers that didn't understand CSS and thus treated `style` as an unknown element.  This is also irrelevant for modern times, and so it's fine to not support it.

So all in all, I'm fine with switching to this algorithm.

Comment (bzbarsky) [reply index.bs]: > and then we should force that here

I'm not following.   The input here is a DOMString, not bytes.  What does it mean to parse that "as utf-8"?  Are we expecting people to pass in strings or bytes here?  Is there value in having this function take USVString, or do we not have a CSS parsing  algorithm that takes a Unicode string as input?   If we're supposed to be passing bytes, we should make that explicit by making this a ByteString, but I expect that's not what people actually want here.

Comment (tabatkins) [reply index.bs]: Ah, so, the input to the tokenizer is a stream of unicode codepoints (per <https://drafts.csswg.org/css-syntax/#tokenization>).  The spec also specifies (in [3.2](https://drafts.csswg.org/css-syntax/#input-byte-stream)) that if you're actually starting from a stream of bytes, there are some rules for how to transform it into a stream of codepoints.  (This series of transformations is more explicitly laid out in one spot in [5.3](https://drafts.csswg.org/css-syntax/#parser-entry-points)

That all said, DOMString doesn't actually contain codepoints. :/

Comment (bzbarsky) [reply index.bs]: Does the tokenize take a stream of codepoints or scalar values?

Treating DOMString as a stream of codepoints is doable.   But if you want scalar values, and I suspect you do, USVString is probably a better fit.

Comment (tabatkins) [reply index.bs]: Just codepoints, tho I suspect I did actually mean scalar values. (Back when I wrote this I don't think we were making the distinction clearly in DOM.) The fact that you can't use an escape to produce a surrogate suggests that I'm assuming no lone surrogates would show up in the input stream either.

File: index.bs (hunk 1 above)
Comment (domenic) [reply index.bs]: I see. As @bzbarsky says, a bit annoying to have to track that. Oh well.

It seems like in the setter for adoptedStyleSheets, you need to loop over all current adopted stylesheets and remove the relevant document from that stylesheets list of adopter documents, to keep it accurate in cases like 

```js
document.adoptedStyleSheets = [s1];
document.adoptedStyleSheets = [];
```

I'd also suggest adding a line like

```html
1. Let |adopterDocument| be this {{DocumentOrShadowRoot}}, if this {{DocumentOrShadowRoot}} is a {{Document}}, or this {{DocumentOrShadowRoot}}'s [=node document=], if this {{DocumentOrShadowRoot}} is a {{ShadowRoot}}.
```

File: index.bs (hunk 2 above)
Comment (domenic) [reply index.bs]: I see. I think it's automatic in the spec too. I would turn it into a non-normative note, and rephrase it to be clearer with something like

```html
<p class="note">
Because the [=adopted stylesheets=] are a property of the {{DocumentOrShadowRoot}}, they move along with the {{ShadowRoot}} if it gets [=adopted=] into a different {{Document}}, e.g. when adopting its [=shadow host=].
</p>
```

I guess I'm still not sure what @bzbarsky was concerned about in #46 though, so maybe he can chime in.

File: index.bs (hunk 3 above)
Comment (domenic) [reply index.bs]: Keeping the sentences is good, but you should change them to say "will throw an exception" instead of "must throw an exception". "Must" is a requirement that implementers should implement, which is confusing because then you have the requirement in two places: both that sentence, and the algorithms themselves. Whereas "will" implies that it will happen because some other normative text, i.e. the below algorithms.

You could even say "will throw an exception, as detailed in the below algorithms".

A nit: you used "this sheet" in step 3 but _sheet_ in steps 1 and 2.

File: index.bs (hunk 1 above)
Comment (rakina) [reply index.bs]: If we want to track removal as well, we'd have to make it a multiset (or make it be a pair of <Document, count>). I guess the multiset thing is easier to have in spec, is that OK?

Comment (domenic) [reply index.bs]: Yeah, I guess it would just be a [list](https://infra.spec.whatwg.org/#lists) then (instead of a set). Just be careful when removing to only remove one.

File: index.bs (hunk 3 above)
Comment (rakina) [reply index.bs]: Done!

File: index.bs (hunk 2 above)
Comment (rakina) [reply index.bs]: Changed that into a note, does this cover your concern well @bzbarsky?

File: index.bs (hunk 4 above)
Comment (rakina) [reply index.bs]: I don't really understand the conversation, but I take that it's OK to use the "parse list of rules" because the charset and comments thing are legacy and OK to not be supported?

Comment (tabatkins) [reply index.bs]: Yes. There might be a little detail I need to fix up in Syntax to handle (filed as <https://github.com/w3c/csswg-drafts/issues/3307>), but that either won't need you to do anything, or will just require a small addition that I'll ping you about.

Comment (rakina) [reply index.bs]: Cool! Do you think the rest of this PR is ok to merge?
